### ISSUE TITLE
bash eval set command not producing expected output in gramine-sgx

### ISSUE THREAD

**Issue by amathew3** [gramineproject/gramine#1671]
### Description of the problem

One of the workload startup script(shell) use `eval set` command to form the argument list. Observed that `gramine-sgx` is not producing the expected output, `gramine-direct` the output is proper and expected one. 

### Steps to reproduce

Run the below sample code.

```
#!/usr/bin/env bash
PRG="$0"
echo $PRG
save () {
    for i do printf %s\\n "$i" | sed "s/'/'\\\\''/g;1s/^/'/;\$s/\$/' \\\\/" ; done
    echo " "
}
APP_ARGS=`save "$@"`
echo "app_args = $APP_ARGS"
eval set -- "$APP_ARGS"
echo $@
```
Native output  of the script.
````
./test_eval --config-file=mysample.toml
./test_eval
app_args = '--config-file=mysample.toml' \

--config-file=mysample.toml
`````
gramine-sgx output
```
gramine-sgx test_eval --config-file=mysample.toml
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - sgx.debug = true                           (this is a debug enclave)
  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)
  - loader.insecure__use_host_env = true       (forwarding environment vars from untrusted host to the app)
  - sys.insecure__allow_eventfd = true         (host-based eventfd is enabled)
  - sgx.file_check_policy = allow_all_but_log  (all files are passed through from untrusted host without verification)
  - sgx.allowed_files = [ ... ]                (some files are passed through from untrusted host without verification)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

test_eval

app_args = '--config-file=mysample.toml' \
--config-file=mysample.toml \
```
the `\` is not removed when executed in gramine-sgx.

### Expected results

output should remove the  escape character and spaces.

### Actual results

Escape character not removed `gramine-sgx` output

### Gramine commit hash

fa437cbcf380ea9c1aa76958ddeb73557a344901

**Comment by dimakuv** [comment]
I can confirm that there's a spurious `\` symbol when executing with `gramine-sgx`. Below is my reproducer:

- Preparation:
```bash
gramine/CI-Examples/bash$ ls -lah
-rwxrwxr-x  ... test_eval

gramine/CI-Examples/bash$ cat test_eval
#!/usr/bin/env bash
PRG="$0"
echo $PRG
save () {
    for i do printf %s\\n "$i" | sed "s/'/'\\\\''/g;1s/^/'/;\$s/\$/' \\\\/" ; done
    echo " "
}
APP_ARGS=`save "$@"`
echo "app_args = $APP_ARGS"
eval set -- "$APP_ARGS"
echo $@
```

- Bash example's template file is modified like this:
```diff
diff --git a/CI-Examples/bash/manifest.template b/CI-Examples/bash/manifest.template
index 9757f5b6..c993451b 100644
--- a/CI-Examples/bash/manifest.template
+++ b/CI-Examples/bash/manifest.template
@@ -2,12 +2,12 @@
 # including ls, cat, cp, date, and rm.

 loader.entrypoint = "file:{{ gramine.libos }}"
-libos.entrypoint = "{{ execdir }}/bash"
+libos.entrypoint = "/test_eval"

 loader.log_level = "{{ log_level }}"

 loader.env.LD_LIBRARY_PATH = "/lib:{{ arch_libdir }}:/usr{{ arch_libdir }}"
-loader.env.PATH = "{{ execdir }}"
+loader.env.PATH = "{{ execdir }}:/usr/{{ execdir }}"

 loader.insecure__use_cmdline_argv = true

@@ -16,6 +16,8 @@ fs.mounts = [
   { path = "{{ arch_libdir }}", uri = "file:{{ arch_libdir }}" },
   { path = "/usr/lib", uri = "file:/usr/lib" },
   { path = "{{ execdir }}", uri = "file:{{ execdir }}" },
+  { path = "/usr/{{ execdir }}", uri = "file:/usr/{{ execdir }}" },
+  { path = "/test_eval", uri = "file:test_eval" },
 ]

 sgx.debug = true
@@ -26,9 +28,11 @@ sgx.max_threads = {{ '1' if env.get('EDMM', '0') == '1' else '4' }}
 sgx.trusted_files = [
   "file:{{ gramine.libos }}",
   "file:{{ execdir }}/",
+  "file:/usr/{{ execdir }}/",
   "file:{{ gramine.runtimedir() }}/",
   "file:{{ arch_libdir }}/",
   "file:/usr/{{ arch_libdir }}/",
+  "file:test_eval",
 ]

 sgx.allowed_files = [
```

- Runs:
```bash
gramine/CI-Examples/bash$ ./test_eval --config-file=mysample.toml
./test_eval
app_args = '--config-file=mysample.toml' \

--config-file=mysample.toml

gramine/CI-Examples/bash$ gramine-direct ./bash --config-file=mysample.toml
/test_eval
app_args = '--config-file=mysample.toml' \

--config-file=mysample.toml

gramine/CI-Examples/bash$ gramine-sgx ./bash --config-file=mysample.toml
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - sgx.debug = true                           (this is a debug enclave)
  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)
  - sgx.allowed_files = [ ... ]                (some files are passed through from untrusted host without verification)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

/test_eval
app_args = '--config-file=mysample.toml' \
--config-file=mysample.toml \
```

- Tested on Git commit: fa437cbcf380ea9c1aa76958ddeb73557a344901

- Ubuntu 20.04

## Quick thoughts

I see that 4 processes are created:
1. Bash parent
2. Bash child
3. Bash grandchild (not sure what this one does, looks like prints the first `/test_eval` string)
4. Sed (created by 2)

**Comment by dimakuv** [comment]
Looking a bit more into this, simplified the bash script and found the culprit:
```bash
gramine/CI-Examples/bash$ cat test_eval
#!/usr/bin/env bash
save () {
    for i do printf %s\\n "$i" | sed "s/'/'\\\\''/g;1s/^/'/;\$s/\$/' \\\\/" ; done
    echo " "  # <-- REMOVING THIS KEEPS THE TRAILING / SYMBOL, EVEN IN NATIVE
}
APP_ARGS=`save "$@"`
eval set -- "$APP_ARGS"
echo $@
```

**Comment by dimakuv** [comment]
I analyzed this issue, and it's a very subtle problem with pipes.

Here is the relevant snippet in Gramine-SGX log:
```
(libos_parser.c:1658:buf_write_all) [P4:T34:sed] trace: ---- write(1, 0x154054d0, 0x20) ...
(libos_parser.c:1658:buf_write_all) [P4:T34:sed] trace: ---- return from write(...) = 0x20
(libos_parser.c:1658:buf_write_all) [P1:T1:bash] trace: ---- return from read(...) = 0x20

(libos_parser.c:1658:buf_write_all) [P1:T1:bash] trace: ---- read(3, 0x11957680, 0x80) ...
(libos_parser.c:1658:buf_write_all) [P2:T2:bash] trace: ---- write(1, 0x13ac02e0, 0x2) ...
(libos_parser.c:1658:buf_write_all) [P2:T2:bash] trace: ---- return from write(...) = 0x2
(libos_parser.c:1658:buf_write_all) [P1:T1:bash] trace: ---- return from read(...) = -13
```

Recall that each pipe in Gramine-SGX is encrypted using TLS. TLS sessions can be setup **only between two peers**, so our encrypted pipes do not support a possible case when **more than two processes use the same pipe**.

This is exactly the case that happens above -- the same pipe is shared between P1, P2 and P4. P4 starts the TLS communication with P1 first, so P1 "binds" the TLS session to P4 and only to P4. Then, if P1 sees a TLS message from P2 (on the same pipe), it considers it a security violation and stops execution.

I added logs, and indeed I observed that `mbedtls_ssl_read()` returns `-0x7180` (aka `MBEDTLS_ERR_SSL_INVALID_MAC`). Which means that P1 **rejects the message** from P2. And this message is the `echo " "` result, see line 4 in the test bash script.

---

How to fix that? Unfortunately I don't see a fix for this inside Gramine. We hit a fundamental limitation of the TLS protocol. TLS is simply **at odds** with sharing pipes between 3+ entities.

Interestingly, this is the first time we observe such behavior in smth like 3 years of Gramine usage. When implementing these TLS-encrypted pipes, I assumed that no sane workload would ever want to share a pipe. Apparently I was wrong.
